neffint/adaptive_fourier_integral.py

```diff
@@ -0,0 +1,466 @@
+#   Copyright 2023 neffint
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import logging
+from typing import Callable, Dict, Optional, Sequence, Tuple
+
+import numpy as np
+import sortednp
+from numpy.typing import ArrayLike
+
+from .fixed_grid_fourier_integral import fourier_integral_fixed_sampling
+from .utils import complex_pchip
+
+MAX_FREQUENCY = 1e60
+
+class CachedFunc:
+    """Wrapper class around a function to cache function calls for faster evaluation of the function with the same arguments.
+    """
+    
+    def __init__(self, function: Callable[[float], ArrayLike]):
+        """Create a wrapper around the input function that saves function calls in a dictionary for faster reevaluation with the same argument.
+
+        :param function: The function to be wrapped. Must be a function of one parameter, which must be hashable.
+        :type function: Callable[[float], ArrayLike]
+        """
+        
+        self.function = function
+        self.cache: Dict[float, ArrayLike] = dict()
+    
+    def __call__(self, x: float) -> ArrayLike:
+        """Evaluate func
+        
+        Defaults to looking up the input variable x in the dictionary of cached calls,
+        then evaluates the function itself if no cached output was stored for that input.
+
+        :param x: The input variable of the wrapped function
+        :type x: float
+        :return: The outputs of the wrapped function for the given input
+        :rtype: ArrayLike
+        """
+        if x not in self.cache.keys():
+            self.cache[x] = self.function(x)
+        return self.cache[x]
```

Comment:
Could consider adding methods to `CachedFunc` to explicitly set function values, e.g. to set a limiting value when the frequency approaches 0 or infinity. This would either entail giving the user explicit access to the `CachedFunc` object or passing in the explicit input-output-pairs as e.g. a list of tuples or similar.

It makes the code more complicated, though, and in many settings one might as well set a very low non-zero frequency bound instead.

Reply:
Not sure either, I'll make an issue for it for now